base/shell/explorer/notifyiconscust.cpp

```diff
@@ -20,26 +20,188 @@
+        return FALSE;
+    }
+
+    const T* GetData()
+    {
+        return m_lpData ? (T*)m_lpData : NULL;
+    }
+};
+
+class TrayIconView : public CListView
+{
+    HIMAGELIST hImageList;
+    HWND hCombo;
+    HWND hwndParent;
+    HWND FindTrayToolbarWindowHandle();
+public:
+    ~TrayIconView(){ImageList_Destroy(hImageList);}
+    HWND Create(HWND hwndParent, DWORD id);
+    BOOL AddColumn(INT Index, INT Width, DWORD strId, INT fmt = LVCFMT_CENTER);
+    BOOL InsertItem(LPWSTR data, INT iSubItem, INT id);
+    VOID BuildIconList();
+};
+
+
+HWND TrayIconView::FindTrayToolbarWindowHandle()
+{
+    HWND hwnd;
+    hwnd = FindWindowW(L"Shell_TrayWnd", NULL);
+    hwnd = FindWindowExW(hwnd, NULL, L"TrayNotifyWnd", NULL);
+    hwnd = FindWindowExW(hwnd, NULL, L"SysPager", NULL);
+    hwnd = FindWindowExW(hwnd, NULL, L"ToolbarWindow32", NULL);
+    return hwnd;
+}
+
+HWND TrayIconView::Create(HWND hwndParent, DWORD id)
+{
+    m_hWnd = ::GetDlgItem(hwndParent, id);
+    hCombo = ::GetDlgItem(hwndParent, IDC_NOTIFICATION_BEHAVIOUR);
+
+    SetExtendedListViewStyle(LVS_EX_FULLROWSELECT, LVS_EX_FULLROWSELECT);
+
+    AddColumn(0, 146, IDS_NOTIFY_COLUMN1);
+    AddColumn(1, 180, IDS_NOTIFY_COLUMN2);
+
+    for (int i = 0; i < 3; i++)
+    {
+        CStringW szNotify(MAKEINTRESOURCEW(IDS_NOTIFY_SHOWALL + i));
+        SendMessageW(hCombo, CB_ADDSTRING, i, (LPARAM)szNotify.GetString());
+    }
+    return  m_hWnd;
+}
+
+BOOL TrayIconView::AddColumn(INT Index, INT Width, DWORD strId, INT fmt)
+{
+    CStringW strLabel(MAKEINTRESOURCEW(strId));
+    LVCOLUMNW Column;
+    ZeroMemory(&Column, sizeof(Column));
+
+    Column.mask =  LVCF_WIDTH | LVCF_SUBITEM | LVCF_TEXT | LVCF_FMT;
+    Column.iSubItem = Index;
+    Column.cx = Width;
+    Column.pszText = (LPWSTR)strLabel.GetString();
+    Column.cchTextMax = strLabel.GetLength();
+    Column.fmt = fmt;
+
+    return (InsertColumn(Index, &Column) == -1) ? FALSE : TRUE;
+}
+
+BOOL TrayIconView::InsertItem(LPWSTR data, INT iSubItem, INT id)
+{
+    LVITEMW iItem;
+
+    iItem.mask = LVIF_TEXT | LVIF_IMAGE;
+    iItem.pszText = data;
+    iItem.cchTextMax = lstrlenW(data);
+    iItem.iSubItem = iSubItem;
+    iItem.iItem = id;
+    iItem.iImage = id;
+
+    if (iSubItem)
+        return (ListView_SetItem(m_hWnd, &iItem) != -1);
+    else
+        return (ListView_InsertItem(m_hWnd, &iItem) != -1);
+}
+
+VOID TrayIconView::BuildIconList()
+{
+    HWND hToolbar = FindTrayToolbarWindowHandle();
+    int count = SendMessageW(hToolbar, TB_BUTTONCOUNT, 0, 0);
+
+    hImageList = ImageList_Create(16, 16, ILC_COLOR24, count, 0);
+
+    TrayIconData<TBBUTTON> IconData(LMEM_FIXED);
+    TBBUTTON tb = {0};
+    NOTIFYICONDATAW idata;
+
+    for(int idx =0 ; idx < count; idx++)
+    {
+        ::SendMessageW(hToolbar, TB_GETBUTTON, idx, (LPARAM)IconData.GetData());
+        IconData.ReadData(&tb);
+        IconData.ReadData<NOTIFYICONDATAW>(&idata,(PCHAR)tb.dwData);
+ 
+        HICON hIcon = CopyIcon(idata.hIcon);
```

Comment:
This should be more or less equivalent, without some extra copies and the `TrayIconData` object.
```
TBBUTTON tb = {0};

for(int idx = 0; idx < count; idx++)
{
    ::SendMessageW(hToolbar, TB_GETBUTTON, idx, (LPARAM)&tb);
    NOTIFYICONDATAW* idata = (NOTIFYICONDATAW*)tb.dwData;
    HICON hIcon = CopyIcon(idata->hIcon);
```